test/framework/src/main/java/org/elasticsearch/common/util/concurrent/DeterministicTaskQueue.java

```diff
@@ -262,278 +264,345 @@ public ThreadPool getThreadPool() {
-                    @Override
-                    public long getDelay(TimeUnit unit) {
-                        throw new UnsupportedOperationException();
-                    }
-
-                    @Override
-                    public int compareTo(Delayed o) {
-                        throw new UnsupportedOperationException();
-                    }
-
-                    @Override
-                    public boolean cancel() {
-                        return taskState.compareAndSet(NOT_STARTED, CANCELLED);
-                    }
-
-                    @Override
-                    public boolean isCancelled() {
-                        return taskState.get() == CANCELLED;
-                    }
-
-                };
+            public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) {
+                throw new UnsupportedOperationException();
             }
 
             @Override
-            public void shutdown() {
+            public <T> T invokeAny(Collection<? extends Callable<T>> tasks) {
                 throw new UnsupportedOperationException();
             }
 
             @Override
-            public void shutdownNow() {
+            public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) {
                 throw new UnsupportedOperationException();
             }
 
             @Override
-            public boolean awaitTermination(long timeout, TimeUnit unit) {
-                throw new UnsupportedOperationException();
+            public void execute(Runnable command) {
+                scheduleNow(runnableWrapper.apply(command));
             }
 
             @Override
-            public ScheduledExecutorService scheduler() {
-                return new ScheduledExecutorService() {
-                    @Override
-                    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
-                        throw new UnsupportedOperationException();
-                    }
+            public String toString() {
+                return "DeterministicTaskQueue/forkingExecutor";
+            }
+        };
 
-                    @Override
-                    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
-                        throw new UnsupportedOperationException();
-                    }
+        protected final Function<Runnable, Runnable> runnableWrapper;
 
-                    @Override
-                    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit) {
-                        throw new UnsupportedOperationException();
-                    }
+        protected DeterministicThreadPool(Function<Runnable, Runnable> runnableWrapper) {
+            this.runnableWrapper = runnable -> getThreadContext().preserveContext(runnableWrapper.apply(runnable));
```

Comment:
This is the fix for always preserving threadContext for each task. Also added a test for it in 2ad396563f13d07a61a834adc34969cd5f2e5549